src/Dialect/ONNX/Transforms/Decompose.cpp

```diff
@@ -333,6 +333,122 @@ bool hasStaticSpatialDims(Value v) {
   return llvm::none_of(Ds, ShapedType::isDynamic);
 }
 
+// In the following pattern, a SequenceAt can be replaced with Split
+//   %seq = onnx.SplitToSequence(%input, %split) {%axis : }
+//   %res = onnx.SequenceAt(%seq, %position)
+// We just try to avoid using the sequence related ops, which are less
+// optimized, or even not implemented in onnx-mlir.
+// In the targeted use case, %split and %position are constant scalar and the
+// tensor of %input and %res have static shape.
+// This condition greatly reduces the complexity of code generation to replace
+// SequenceAt with split op
+//   %res = onnx.Split(%input, onnx.expand(%split, %input.shape()[%axis]))
+//   {%axis : } : %position
+// onnx.expand(%split, %input.shape()[%axis]) can be a constant under the
+// assumed condition.
+// Here %position has to be compiler time constant.
+// For multiple SequenceAt from the same SplitToSequence result, the onnx.split
+// for different SequenceAt are expected to be merged by optimization.
+// Alternatively, Slice can be used
+//   %res = onnx.Slice(%input, %start, %end, %step)
+// The start, and end for slice will be onnx.constant:
+//   start: %position*%split for %axis, 0 for other dimensionis
+//   end: (%positiion+1)*%split for %axis, upper bound for other dimension
+//   step: 1 for all dimensions
+// The split approach may have better performance than the alternative slice
+// approach,  because the slicing is done separately.
+
+bool canSequenceAtBeReplaced(Value sequenceAtResult) {
+  if (!hasStaticShape(sequenceAtResult.getType()))
+    return false;
+
+  ONNXSequenceAtOp op =
+      mlir::dyn_cast<ONNXSequenceAtOp>(sequenceAtResult.getDefiningOp());
+
+  Value inputSequence = op.getInputSequence();
+  Value position = op.getPosition();
+
+  if (!isDenseONNXConstant(position))
+    return false;
+
+  // Input sequence should be defined with SplitToSequence
+  ONNXSplitToSequenceOp splitToSequence;
+  if (!(splitToSequence = mlir::dyn_cast<ONNXSplitToSequenceOp>(
+            inputSequence.getDefiningOp())))
+    return false;
+
+  // Check the pattern of the SplitToSequence op
+  Value input = splitToSequence.getInput();
+  if (!hasStaticShape(input.getType()))
+    return false;
+  Value split = splitToSequence.getSplit();
+  if (!isScalarConstantTensor(split))
+    return false;
+
+  return true;
+}
+
+Value replaceSequenceAt(
+    PatternRewriter &rewriter, Location loc, Value sequenceAtResult) {
+  ONNXSequenceAtOp op =
+      mlir::cast<ONNXSequenceAtOp>(sequenceAtResult.getDefiningOp());
```

Comment:
Same here, we can use `ONNXSequenceAtOp op = sequenceAtResult.getDefiningOp<ONNXSequenceAtOp>()`